espnet2/layers/augmentation.py

```diff
@@ -0,0 +1,535 @@
+
+    Returns:
+        ret (torch.Tensor): filtered signal (..., time)
+    """
+    ret = torchaudio.functional.equalizer_biquad(
+        waveform, sample_rate, center_freq, gain, Q=Q
+    )
+    return ret
+
+
+def pitch_shift(
+    waveform,
+    sample_rate: int,
+    n_steps: int,
+    bins_per_octave: int = 12,
+    n_fft: float = 0.032,
+    win_length: Optional[float] = None,
+    hop_length: float = 0.008,
+    window: Optional[str] = "hann",
+):
+    """Shift the pitch of a waveform by `n_steps` steps.
+
+    Note: this function is slow.
+
+    Args:
+        waveform (torch.Tensor): audio signal (..., time)
+        sample_rate (int): sampling rate in Hz
+        n_steps (int): the (fractional) steps to shift the pitch
+            -4 for shifting pitch down by 4/`bins_per_octave` octaves
+            4 for shifting pitch up by 4/`bins_per_octave` octaves
+        bins_per_octave (int): number of steps per octave
+        n_fft (int): length of FFT (in second)
+        win_length (int or None): The window length (in second) used for STFT
+            If None, it is treated as equal to n_fft
+        hop_length (int): The hop size (in second) used for STFT
+        window (str or None): The windowing function applied to the signal after
+            padding with zeros
+
+    Returns:
+        ret (torch.Tensor): filtered signal (..., time)
+    """
+    n_fft = int(sample_rate * n_fft)
+    if hop_length is None:
+        hop_length = n_fft // 4
+    else:
+        hop_length = int(sample_rate * hop_length)
+    if win_length is None:
+        win_length = n_fft
+    if window is not None:
+        window_func = getattr(torch, f"{window}_window")
+        window = window_func(win_length, dtype=waveform.dtype, device=waveform.device)
+    ret = torchaudio.functional.pitch_shift(
+        waveform,
+        sample_rate,
+        n_steps,
+        bins_per_octave=bins_per_octave,
+        n_fft=n_fft,
+        win_length=win_length,
+        hop_length=hop_length,
+        window=window,
+    )
+    return ret
+
+
+def speed_perturb(waveform, sample_rate: int, factor: float):
+    """Speed perturbation which also changes the pitch.
+
+    Note: This function should be used with caution as it changes the signal duration.
+
+    Args:
+        waveform (torch.Tensor): audio signal (..., time)
+        sample_rate (int): sampling rate in Hz
+        factor (float): speed factor (e.g., 0.9 for 90% speed)
+        lengths (torch.Tensor): lengths of the input signals
+
+    Returns:
+        ret (torch.Tensor): perturbed signal (..., time)
+    """
+    orig_freq = sample_rate
+    source_sample_rate = int(factor * orig_freq)
+    target_sample_rate = int(orig_freq)
+
+    gcd = math.gcd(source_sample_rate, target_sample_rate)
+    source_sample_rate = source_sample_rate // gcd
+    target_sample_rate = target_sample_rate // gcd
+
+    ret = torchaudio.functional.resample(
```

Comment:
Did you consider applying one without pitch shift? Would it cause severe more computation? 
Also, how's the training training speed with this augment (bottlenecks in loading)?

Reply:
`speed_perturb` and `time_stretch` are two different time scaling method. The former changes the pitch while the latter does not. I think it is dependent on the use case. So I just provide both for the user to choose.

I haven't strictly tested the speed difference yet. Will do some test later.